### ISSUE TITLE
Packaging issue with Tensorflow's _pywrap_tensorflow_internal 

### ISSUE THREAD

**Issue by Overdrivr** [pyinstaller/pyinstaller#3754]
There are several issues on SO (like [this one](https://stackoverflow.com/questions/49620638/tensorflow-gpu-with-pyinstaller)) regarding executables generated by pyinstaller that require tensorflow .

On program startup, the following error happens:
```
module `tensorflow.python._pywrap_tensorflow_internal` could not be found.
```

I am facing the same issue, for some reason only with `tensorflow-gpu` version, I have not witnessed this problem on cpu-based `tensorflow` module.

My spec file does not do anything special, except declaring it as `hiddenimport` because I import TF from a cythonized python package.

The problem is rather straightforward and can be solved manually. By looking at pyinstaller's generated folder, it appears the `_pywrap_tensorflow_internal` is there, but at the root directory instead of `tensorflow/python/` as it would be expected:

![image](https://user-images.githubusercontent.com/1294805/45828423-c94d5580-bcf8-11e8-828f-a6bbf90d228c.png)
(sorry about the artistic skills)

Moving the file manually to a newly created `tensorflow/python/` folder solves the issue.

I'm not familiar with pyinstaller's hooks, do you think it's something that could be solved as part of one ? In that case I would gladly contribute and help fix this.

**Comment by htgoebel** [comment]
Yes, this looks like a typical case for a hook.

I'm wonder where this `_pywrap_tensorflow_internal` directory comes from, though, since there is no hook for tensorflow yet. What does this directory contain?

**Comment by Overdrivr** [comment]
Hum, looks like the issue is more serious than expected. 
Moving the file seems to break something as soon as I move the entire executable to another (exactly identical) Windows 10 machine.

Initially, I was getting `DLL load failed, module not found error` regarding `_pywrap_tensorflow_internal.pyd`.

So I modified my spec file as such, to rename this file to the path where it seems to be expected.

```
              'tensorflow',
              'tensorflow.python._pywrap_tensorflow_internal'
             ],
             hookspath=[],
             runtime_hooks=[],
             excludes=[],
             win_no_prefer_redirects=False,
             win_private_assemblies=False,
             cipher=block_cipher)
pyz = PYZ(a.pure, a.zipped_data,
             cipher=block_cipher)

# Rename _pywrap_tensorflow_internal to tensorflow.python._pywrap_tensorflow_internal
for i in range(len(a.binaries)):
  dest, origin, kind = a.binaries[i]

  if '_pywrap_tensorflow_internal' in dest:
      print('Remapping TF-INTERNAL')
      a.binaries[i] = ('tensorflow.python.' + dest, origin, kind)
      import pdb; pdb.set_trace()

exe = EXE(pyz,
...
```

Now I'm getting another error

```
Traceback (most recent call last):
  File "site-packages\tensorflow\python\pywrap_tensorflow.py", line 58, in <module>
  File "c:\users\paperspace\.virtualenvs\****-ecouzwts\lib\site-packages\PyInstaller\loader\pyimod03_importers.py", line 627, in exec_module
  File "site-packages\tensorflow\python\pywrap_tensorflow_internal.py", line 18, in <module>
  File "site-packages\tensorflow\python\pywrap_tensorflow_internal.py", line 17, in swig_import_helper
  File "importlib\__init__.py", line 126, in import_module
ImportError: DLL load failed: A dynamic link library (DLL) initialization routine failed.


Failed to load the native TensorFlow runtime.

See https://www.tensorflow.org/install/install_sources#common_installation_problems

for some common reasons and solutions.  Include the entire stack trace
above this error message when asking for help.
```

To summarize:
- I have the issue with Tensorflow-GPU versions only
- Executable generated by PyInstaller on one Win10 machine does not work on another Win10 machine (they are strictly identical os-wise, booted with paperspace)

Has anyone succeeded in packaging a tensorflow application using pyinstaller ?

**Comment by Overdrivr** [comment]
> I'm wonder where this _pywrap_tensorflow_internal directory comes from, though, since there is no hook for tensorflow yet. What does this directory contain?

It's not a directory, it's a binary, I have the impression that the entire tensorflow runtime is packaged in this .pyd (this is the only TF-related file in my pyinstaller-generated executable)

**Comment by Overdrivr** [comment]
Nevermind the issue above. It was actually due a setup issue on the build machine. 

I can add the following code to my `.spec` file, but I am not sure how to do this from a hook (basically, change the location of the binary file).

```
            win_private_assemblies=False,
             cipher=block_cipher)

# custom code to move _pywrap_tensorflow_internal.pyd to tensorflow/python/_pywrap_tensorflow_internal.pyd			 
for i in range(len(a.binaries)):
    dest, origin, kind = a.binaries[i]
    if '_pywrap_tensorflow_internal' in dest:
        a.binaries[i] = ('tensorflow.python.' + dest, origin, kind)

pyz = PYZ(a.pure, a.zipped_data,
             cipher=block_cipher)
```

Is it possible ?

**Comment by songxh2** [comment]
this problem still exist！had it resovled？

**Comment by Overdrivr** [comment]
Well the code of my previous commands does the trick, I've used it on 2 different applications that ship with TF without problem.

One of them is open-source so you can check the whole spec file here: https://gitlab.com/Overdrivr/autosync/blob/master/autosync.spec

**Comment by songxh2** [comment]
> Well the code of my previous commands does the trick, I've used it on 2 different applications that ship with TF without problem.
> 
> One of them is open-source so you can check the whole spec file here: https://gitlab.com/Overdrivr/autosync/blob/master/autosync.spec

`E:\test>label_image_v1.exe
Traceback (most recent call last):
  File "lib\site-packages\tensorflow\python\pywrap_tensorflow.py", line 58, in <module>
  File "f:\pycharmprojects\multi_label_classification\venv\lib\site-packages\PyInstaller\loader\pyimod03_importers.py", line 627, in exec_module
  File "lib\site-packages\tensorflow\python\pywrap_tensorflow_internal.py", line 28, in <module>
  File "lib\site-packages\tensorflow\python\pywrap_tensorflow_internal.py", line 24, in swig_import_helper
  File "imp.py", line 243, in load_module
  File "imp.py", line 343, in load_dynamic
ImportError: DLL load failed: 动态链接库(DLL)初始化例程失败。

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "label_image_v1.py", line 13, in <module>
  File "f:\pycharmprojects\multi_label_classification\venv\lib\site-packages\PyInstaller\loader\pyimod03_importers.py", line 627, in exec_module
  File "lib\site-packages\tensorflow\__init__.py", line 24, in <module>
  File "f:\pycharmprojects\multi_label_classification\venv\lib\site-packages\PyInstaller\loader\pyimod03_importers.py", line 627, in exec_module
  File "lib\site-packages\tensorflow\python\__init__.py", line 49, in <module>
  File "f:\pycharmprojects\multi_label_classification\venv\lib\site-packages\PyInstaller\loader\pyimod03_importers.py", line 627, in exec_module
  File "lib\site-packages\tensorflow\python\pywrap_tensorflow.py", line 74, in <module>
ImportError: Traceback (most recent call last):
  File "lib\site-packages\tensorflow\python\pywrap_tensorflow.py", line 58, in <module>
  File "f:\pycharmprojects\multi_label_classification\venv\lib\site-packages\PyInstaller\loader\pyimod03_importers.py", line 627, in exec_module
  File "lib\site-packages\tensorflow\python\pywrap_tensorflow_internal.py", line 28, in <module>
  File "lib\site-packages\tensorflow\python\pywrap_tensorflow_internal.py", line 24, in swig_import_helper
  File "imp.py", line 243, in load_module
  File "imp.py", line 343, in load_dynamic
ImportError: DLL load failed: 动态链接库(DLL)初始化例程失败。


Failed to load the native TensorFlow runtime.

See https://www.tensorflow.org/install/errors

for some common reasons and solutions.  Include the entire stack trace
above this error message when asking for help.
[15164] Failed to execute script label_image_v1`

**Comment by songxh2** [comment]
> > Well the code of my previous commands does the trick, I've used it on 2 different applications that ship with TF without problem.
> > One of them is open-source so you can check the whole spec file here: https://gitlab.com/Overdrivr/autosync/blob/master/autosync.spec
> 
> `E:\test>label_image_v1.exe
> Traceback (most recent call last):
> File "lib\site-packages\tensorflow\python\pywrap_tensorflow.py", line 58, in 
> File "f:\pycharmprojects\multi_label_classification\venv\lib\site-packages\PyInstaller\loader\pyimod03_importers.py", line 627, in exec_module
> File "lib\site-packages\tensorflow\python\pywrap_tensorflow_internal.py", line 28, in 
> File "lib\site-packages\tensorflow\python\pywrap_tensorflow_internal.py", line 24, in swig_import_helper
> File "imp.py", line 243, in load_module
> File "imp.py", line 343, in load_dynamic
> ImportError: DLL load failed: 动态链接库(DLL)初始化例程失败。
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
> File "label_image_v1.py", line 13, in 
> File "f:\pycharmprojects\multi_label_classification\venv\lib\site-packages\PyInstaller\loader\pyimod03_importers.py", line 627, in exec_module
> File "lib\site-packages\tensorflow__init__.py", line 24, in 
> File "f:\pycharmprojects\multi_label_classification\venv\lib\site-packages\PyInstaller\loader\pyimod03_importers.py", line 627, in exec_module
> File "lib\site-packages\tensorflow\python__init__.py", line 49, in 
> File "f:\pycharmprojects\multi_label_classification\venv\lib\site-packages\PyInstaller\loader\pyimod03_importers.py", line 627, in exec_module
> File "lib\site-packages\tensorflow\python\pywrap_tensorflow.py", line 74, in 
> ImportError: Traceback (most recent call last):
> File "lib\site-packages\tensorflow\python\pywrap_tensorflow.py", line 58, in 
> File "f:\pycharmprojects\multi_label_classification\venv\lib\site-packages\PyInstaller\loader\pyimod03_importers.py", line 627, in exec_module
> File "lib\site-packages\tensorflow\python\pywrap_tensorflow_internal.py", line 28, in 
> File "lib\site-packages\tensorflow\python\pywrap_tensorflow_internal.py", line 24, in swig_import_helper
> File "imp.py", line 243, in load_module
> File "imp.py", line 343, in load_dynamic
> ImportError: DLL load failed: 动态链接库(DLL)初始化例程失败。
> 
> Failed to load the native TensorFlow runtime.
> 
> See https://www.tensorflow.org/install/errors
> 
> for some common reasons and solutions. Include the entire stack trace
> above this error message when asking for help.
> [15164] Failed to execute script label_image_v1`

can run in origin computer,but can not run in another one

**Comment by songxh2** [comment]
i use tensorflow 1.13 version, spec file bellow:
`# -*- mode: python -*-

block_cipher = None


a = Analysis(['label_image_v1.py'],
             pathex=['F:\\PycharmProjects\\multi_label_classification'],
             binaries=[],
             datas=[],
             hiddenimports=[],
             hookspath=[],
             runtime_hooks=[],
             excludes=[],
             win_no_prefer_redirects=False,
             win_private_assemblies=False,
             cipher=block_cipher,
             noarchive=False)

# custom code to move _pywrap_tensorflow_internal.pyd to tensorflow/python/_pywrap_tensorflow_internal.pyd
for i in range(len(a.binaries)):
    dest, origin, kind = a.binaries[i]
    if '_pywrap_tensorflow_internal' in dest:
        a.binaries[i] = ('tensorflow.python.' + dest, origin, kind)

pyz = PYZ(a.pure, a.zipped_data,
             cipher=block_cipher)
exe = EXE(pyz,
          a.scripts,
          a.binaries,
          a.zipfiles,
          a.datas,
          [],
          name='label_image_v1',
          debug=False,
          bootloader_ignore_signals=False,
          strip=False,
          upx=True,
          runtime_tmpdir=None,
          console=True )
`

**Comment by songxh2** [comment]
Use tensorflow 1.13
Build platform Win 10 64bit
Running platform Win 10 64bit

**Comment by Overdrivr** [comment]
This error is pretty common, and most likely has nothing to do with pyinstaller. 
TF 1.13 is built with AVX support (from TF 1.6), so if your running platform does not support AVX you'll get this error message. Reply with system information, I'll tell you if it supports AVX or not.

OR, install tensorflow from wheel file directly, using wheels from this repository: https://github.com/fo40225/tensorflow-windows-wheel
I've done it yesterday with TF CPU 1.10 using sse2 wheels, works perfect.

Sometimes this error is also due to missing [Visual C++ 2015 redistributable](https://www.microsoft.com/fr-fr/download/details.aspx?id=48145), but that's more rare.

**Comment by songxh2** [comment]
> This error is pretty common, and most likely has nothing to do with pyinstaller.
> TF 1.13 is built with AVX support (from TF 1.6), so if your running platform does not support AVX you'll get this error message. Reply with system information, I'll tell you if it supports AVX or not.
> 
> OR, install tensorflow from wheel file directly, using wheels from this repository: https://github.com/fo40225/tensorflow-windows-wheel
> I've done it yesterday with TF CPU 1.10 using sse2 wheels, works perfect.
> 
> Sometimes this error is also due to missing [Visual C++ 2015 redistributable](https://www.microsoft.com/fr-fr/download/details.aspx?id=48145), but that's more rare.

thanks,i got it

**Comment by UN-CharlieH** [comment]
Same problem with tensorflow 1.13.1. Works on dev system, fails on VM, even with installed runtimes.
I believe this is the problem:
```
Dump of file _pywrap_tensorflow_internal.pyd

File Type: DLL

  Image has the following dependencies:

    python37.dll
    ADVAPI32.dll
    MSVCP140.dll
    WS2_32.dll
    SHLWAPI.dll
    KERNEL32.dll
    VCRUNTIME140.dll
    api-ms-win-crt-runtime-l1-1-0.dll
    api-ms-win-crt-math-l1-1-0.dll
    api-ms-win-crt-stdio-l1-1-0.dll
    api-ms-win-crt-string-l1-1-0.dll
    api-ms-win-crt-heap-l1-1-0.dll
    api-ms-win-crt-utility-l1-1-0.dll
    api-ms-win-crt-environment-l1-1-0.dll
    api-ms-win-crt-filesystem-l1-1-0.dll
    api-ms-win-crt-convert-l1-1-0.dll
    api-ms-win-crt-time-l1-1-0.dll
```
That is, the api-ms-win dependencies. This is the solution required in tensorflow:
[mincore.lib is incorrect](https://github.com/google/UIforETW/commit/24e15292c8d46dae897bf2c3c03c1b1aa2f88731)

**Comment by rlewkowicz** [comment]
@UN-CharlieH 

How did you dump that file?

**Comment by UN-CharlieH** [comment]
For generating a list of linked dependencies; dumpbin /dependents <binary-name>

However, it turns out that the VM used for testing had an older XenServer version that did not support AVX. HyperV does, so there was the usual scramble to dig for information about VM differences.

This does raise a question.  Is the mincore.lib dependency (encountered a year ago and solved in the link) how Microsoft introduced the AVX support? If so, then MS may have resolved some edge cases so those dependencies are no longer an issue.

**Comment by snehri** [comment]
> Bu hata oldukça yaygındır ve büyük olasılıkla pyinstaller ile ilgisi yoktur.
> TF 1.13, AVX desteği (TF 1.6'dan) ile üretilmiştir, bu nedenle çalışan platformunuz AVX'i desteklemiyorsa bu hata iletisini alırsınız. Sistem bilgilerini yanıtlayın, size AVX'i destekleyip desteklemediğini söyleyeceğim.
> 
> VEYA, bu depodaki tekerlekleri kullanarak doğrudan tekerlek dosyasından tensorflow'u yükleyin: https://github.com/fo40225/tensorflow-windows-wheel
> Dün TFSE 1.10 ile sse2 tekerlekleri kullanarak yaptım, mükemmel çalışıyor.
> 
> Bazen bu hata, [Visual C ++ 2015'in yeniden dağıtılabilir](https://www.microsoft.com/fr-fr/download/details.aspx?id=48145) olmamasından kaynaklanmaktadır , ancak bu daha nadirdir.

I get a similar error. Can you explain how you use the github link?

**Comment by davideTricarico** [comment]
> This error is pretty common, and most likely has nothing to do with pyinstaller.
> TF 1.13 is built with AVX support (from TF 1.6), so if your running platform does not support AVX you'll get this error message. Reply with system information, I'll tell you if it supports AVX or not.
> 
> OR, install tensorflow from wheel file directly, using wheels from this repository: https://github.com/fo40225/tensorflow-windows-wheel
> I've done it yesterday with TF CPU 1.10 using sse2 wheels, works perfect.
> 
> Sometimes this error is also due to missing [Visual C++ 2015 redistributable](https://www.microsoft.com/fr-fr/download/details.aspx?id=48145), but that's more rare.

Hello

About this issue, do I need to install the sse2 tf on the destination machine or on the machine which generates the pyinstaller exe?

Thanks
Davide

**Comment by snehri** [comment]
Hello,
1.) Visual studio 2015-2019 C ++ red .. create.
2.) please run with administrator.

Savas

Samsung Galaxy akıllı telefonumdan gönderildi.


-------- Orijinal mesaj --------
Başlangıç tarihi: Davide <notifications@github.com>
Tarih: 29.07.2020 15:23 (GMT+03:00)
Alıcı: pyinstaller/pyinstaller <pyinstaller@noreply.github.com>
Cc: snehri <s.nehri@hotmail.com>, Comment <comment@noreply.github.com>
Konu: Re: [pyinstaller/pyinstaller] Packaging issue with Tensorflow's _pywrap_tensorflow_internal  (#3754)


This error is pretty common, and most likely has nothing to do with pyinstaller.
TF 1.13 is built with AVX support (from TF 1.6), so if your running platform does not support AVX you'll get this error message. Reply with system information, I'll tell you if it supports AVX or not.

OR, install tensorflow from wheel file directly, using wheels from this repository: https://github.com/fo40225/tensorflow-windows-wheel
I've done it yesterday with TF CPU 1.10 using sse2 wheels, works perfect.

Sometimes this error is also due to missing Visual C++ 2015 redistributable<https://www.microsoft.com/fr-fr/download/details.aspx?id=48145>, but that's more rare.

Hello

About this issue, do I need to install the sse2 tf on the destination machine or on the machine which generates the pyinstaller exe?

Thanks
Davide

**Comment by Overdrivr** [comment]
@davideTricarico I'm not sure if your build machine needs to support SSE2, maybe without some binaries might be missing and not end up in your package. Not sure, but it's pretty easy to fullfill this requirement on build side.

What's 100% sure and kind of more difficult is to make sure every destination machine your binary will get installed to supports SSE2. If your app is distributed to customer, that can be a pretty difficult requirement.

**Comment by davideTricarico** [comment]
> @davideTricarico I'm not sure if your build machine needs to support SSE2, maybe without some binaries might be missing and not end up in your package. Not sure, but it's pretty easy to fullfill this requirement on build side.
> 
> What's 100% sure and kind of more difficult is to make sure every destination machine your binary will get installed to supports SSE2. If your app is distributed to customers, that can be a pretty difficult requirement.

Thanks for the comment

I tried building the exe on a machine with AVX instruction and running it on the one without: it doesn't work
I'm going to try to build the exe without AVX and try again.

**Comment by bhuvanaprakash** [comment]
The `_pywrap_tensorflow_internal` file is not getting created at all. I'm getting the following error when I try to run the exe file.

> no module named tensorflow
